File: SCG-Miage/src/main/java/Miage/sgc/data/dao/IDAO.java
package Miage.sgc.data.dao;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import Miage.sgc.data.object.BaseObject;

public interface IDAO<T extends BaseObject> {
	static String INSERT = "INSERT INTO {0} {1} VALUES {2}";
    static String UPDATE = "UPDATE {0} SET {1} WHERE ID=?";
    static String UPDATE_QUERY = "UPDATE {0} SET {1} WHERE {3}";
    static String DELETE = "DELETE FROM {0} WHERE ID=?";
    static String DELETE_QUERY = "DELETE FROM {0} WHERE {1}";
    static String SELECT = "SELECT * FROM {0} WHERE ID=?";
    static String SELECT_QUERY = "SELECT * FROM {0} WHERE {1}";
    static String SELECTALL = "SELECT * FROM ";

    /**
     * 
     * @param obj l'objet à créer
     * @return True si la création c'est bien passé, false sinon
     */
    boolean create(T obj);

    /**
     * 
     * @param id
     * @return
     */
    T read(String id);
    
    //List<T> readQuery(); TODO
    
    List<T> readAll();

    boolean update(T obj);
    
    //boolean updateQuery(); TODO

    /**
     * 
     * @param id
     * @return
     */
    boolean delete(String id);
    
    //boolean deleteQuery(); TODO
    
    T setObjectResult(ResultSet resultSet) throws SQLException;
}

File: Dedalus.Core.Api/DAO/IDAO.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Core.Api.DAO
{
    public interface IDAO<T> where T: IBaseObject
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="objet"></param>
        /// <returns></returns>
        Task<bool> Create(T objet);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        Task<T> Delete(string id);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="objet"></param>
        /// <param name="id"></param>
        /// <returns></returns>
        Task<bool> Update(T objet, string id);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        Task<T> Read(string id);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        bool Contains(string id);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="nbPerPage"></param>
        /// <returns></returns>
        IList<T> ReadAll(int nbPerPage = 10);
    }
}
